larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+{
+    return m_caloHits;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+int LArHierarchyHelper::RecoHierarchy::Node::GetParticleId() const
+{
+    return m_pdg;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const std::string LArHierarchyHelper::RecoHierarchy::Node::ToString(const std::string &prefix) const
+{
+    std::string str(prefix + "PDG: " + std::to_string(m_pdg) + " Hits: " + std::to_string(m_caloHits.size()) + "\n");
+    for (const Node *pChild : m_children)
+        str += pChild->ToString(prefix + "   ");
+
+    return str;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHierarchyHelper::MCMatches::MCMatches(const MCHierarchy::Node *pMC) :
+    m_pMC{pMC}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArHierarchyHelper::MCMatches::AddRecoMatch(const RecoHierarchy::Node *pReco, const int nSharedHits)
+{
+    m_recoNodes.emplace_back(pReco);
+    m_sharedHits.emplace_back(nSharedHits);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+int LArHierarchyHelper::MCMatches::GetSharedHits(const RecoHierarchy::Node *pReco) const
+{
+    auto iter{std::find(m_recoNodes.begin(), m_recoNodes.end(), pReco)};
+    if (iter == m_recoNodes.end())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+    int index = iter - m_recoNodes.begin();
+
+    return static_cast<int>(m_sharedHits[index]);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float LArHierarchyHelper::MCMatches::GetPurity(const RecoHierarchy::Node *pReco) const
+{
+    auto iter{std::find(m_recoNodes.begin(), m_recoNodes.end(), pReco)};
+    if (iter == m_recoNodes.end())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+    int index = iter - m_recoNodes.begin();
+
+    return m_sharedHits[index] / static_cast<float>(pReco->GetCaloHits().size());
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float LArHierarchyHelper::MCMatches::GetCompleteness(const RecoHierarchy::Node *pReco) const
+{
+    auto iter{std::find(m_recoNodes.begin(), m_recoNodes.end(), pReco)};
+    if (iter == m_recoNodes.end())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+    int index = iter - m_recoNodes.begin();
```

Comment:
Is there some casting between types going on here? Could make explicit. Const too? Looks OK, though.

Reply:
Switched to std::distance